Remotion/Core/Development.Analyzers/RMCORE0001_SafeContextAnalyzer.cs

```diff
@@ -111,6 +123,89 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
         match.CreateDiagnostic.Invoke(ctx, match.methodSymbol);
     }
 
+    private static void AnalyzeParallelInvocationExpression (
+        SyntaxNodeAnalysisContext ctx,
+        IReadOnlyCollection<(IMethodSymbol methodSymbol, Action<SyntaxNodeAnalysisContext, ISymbol> CreateDiagnostic)> methodSymbols,
+        INamedTypeSymbol? safeContextParallelTypeSymbol)
+    {
+      //Check whether the invocation expression should be covered
+      var invocationSymbolInfo = ctx.SemanticModel.GetSymbolInfo(ctx.Node);
+
+      if (invocationSymbolInfo.Symbol == null)
+        return;
+
+      var invocationOriginalDefinitionSymbol = invocationSymbolInfo.Symbol.OriginalDefinition;
+
+      var match = methodSymbols.FirstOrDefault(m => SymbolEqualityComparer.Default.Equals(m.methodSymbol, invocationOriginalDefinitionSymbol));
+      if (match == default)
+        return;
+
+      // Quit if the type symbol is not available
+      if (safeContextParallelTypeSymbol == null)
+      {
+        match.CreateDiagnostic.Invoke(ctx, match.methodSymbol);
+        return;
+      }
+
+      var openSafeContextBoundaryMethodSymbol = safeContextParallelTypeSymbol.GetMembers("OpenSafeContextBoundary").Cast<IMethodSymbol>().First();
+
+      if (UsesSafeContextBoundaryWithLocalDeclarationUsingStatement(ctx, openSafeContextBoundaryMethodSymbol))
+        return;
+
+      if (UsesSafeContextBoundaryWithUsingStatement(ctx, openSafeContextBoundaryMethodSymbol))
+        return;
+
+      match.CreateDiagnostic.Invoke(ctx, match.methodSymbol);
+    }
+
+    private static bool UsesSafeContextBoundaryWithUsingStatement (SyntaxNodeAnalysisContext ctx, IMethodSymbol? openSafeContextBoundaryMethodSymbol)
+    {
+      var usingStatementSyntax = ctx.Node.FirstAncestorOrSelf<UsingStatementSyntax>(
+          usingStatementSyntax =>
+          {
+            var usingExpression = usingStatementSyntax.Expression;
+            if (usingExpression == null)
+              return false;
+
+            var usingExpressionSymbolInfo = ctx.SemanticModel.GetSymbolInfo(usingExpression);
+
+            return SymbolEqualityComparer.Default.Equals(usingExpressionSymbolInfo.Symbol!.OriginalDefinition, openSafeContextBoundaryMethodSymbol);
+          });
+
+      return usingStatementSyntax != default;
+    }
+
+    private static bool UsesSafeContextBoundaryWithLocalDeclarationUsingStatement (SyntaxNodeAnalysisContext ctx, IMethodSymbol? openSafeContextBoundaryMethodSymbol)
+    {
+      var checkedChildrenSet = new HashSet<SyntaxNode>();
+
+      var blockSyntaxWithRequiredUsingDeclaration = ctx.Node
+          .FirstAncestorOrSelf<BlockSyntax>(blockSyntax => blockSyntax
+              .ChildNodes()
+              .Where(
+                  child => !checkedChildrenSet.Contains(child))
+              .Select(
+                  child =>
+                  {
+                    checkedChildrenSet.Add(child);
+                    return child;
+                  })
```

Comment:
Adding to a `HashSet<T>` returns a bool indicating if the value already existed.
```suggestion
              .Where(child => checkedChildrenSet.Add(child))
```